File: dotnet/Pyrolite/Pickle/Objects/ExceptionConstructor.cs
/* part of Pyrolite, by Irmen de Jong (irmen@razorvine.net) */

using System;
using System.Collections.Generic;
using System.Reflection;

namespace Razorvine.Pickle.Objects
{

/// <summary>
/// This creates PythonException instances. 
/// It keeps track of the original Python exception type name as well.
/// </summary>
public class ExceptionConstructor : IObjectConstructor {

	private readonly string pythonExceptionType;
	private readonly Type type;
	
	public ExceptionConstructor(Type type, string module, string name) {
		if(!string.IsNullOrEmpty(module))
			pythonExceptionType = module+"."+name;
		else
			pythonExceptionType = name;
		this.type = type;
	}

	public object construct(object[] args) {
		try {
			if(!string.IsNullOrEmpty(pythonExceptionType)) {
				// put the python exception type somewhere in the message
				if(args==null || args.Length==0) {
					args = new string[] { "["+pythonExceptionType+"]" };
				} else {
					string msg = (string)args[0];
					msg = string.Format("[{0}] {1}", pythonExceptionType, msg);
					args = new string[] {msg};
				}
			}
			object ex = Activator.CreateInstance(this.type, args);
			
			PropertyInfo prop=ex.GetType().GetProperty("PythonExceptionType");
			if(prop!=null) {
				prop.SetValue(ex, pythonExceptionType, null);
			}
			return ex;
		} catch (Exception x) {
			throw new PickleException("problem constructing object",x);
		}
	}
}

}

File: java/src/main/java/net/razorvine/pickle/objects/ExceptionConstructor.java
package net.razorvine.pickle.objects;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;

import net.razorvine.pickle.IObjectConstructor;
import net.razorvine.pickle.PickleException;

/**
 * This creates Python Exception instances. 
 * It keeps track of the original Python exception type name as well.
 * 
 * @author Irmen de Jong (irmen@razorvine.net)
 */
public class ExceptionConstructor implements IObjectConstructor {

	private Class<?> type;
	private String pythonExceptionType;

	public ExceptionConstructor(Class<?> type, String module, String name) {
		if(module!=null)
			pythonExceptionType = module+"."+name;
		else
			pythonExceptionType = name;
		this.type = type;
	}

	public Object construct(Object[] args) {
		try {
			if(pythonExceptionType!=null) {
				// put the python exception type somewhere in the message
				if(args==null || args.length==0) {
					args = new String[] { "["+pythonExceptionType+"]" };
				} else {
					String msg = "["+pythonExceptionType+"] "+(String)args[0];
					args = new String[] {msg};
				}
			}
			Class<?>[] paramtypes = new Class<?>[args.length];
			for (int i = 0; i < args.length; ++i) {
				paramtypes[i] = args[i].getClass();
			}
			Constructor<?> cons = type.getConstructor(paramtypes);
			Object ex = cons.newInstance(args);
			
			try {
				Field prop = ex.getClass().getField("pythonExceptionType");
				prop.set(ex, pythonExceptionType);
			} catch (NoSuchFieldException x) {
				// meh.
			}
			return ex;
		} catch (Exception x) {
			throw new PickleException("problem construction object: " + x);
		}
	}
}
